File: Tests.Csharp/Subscribing.cs
using System.Text.Json;
using UmaDb.Csharp;
using UmaDb.Csharp.Messages;
using Xunit;

namespace Tests.Csharp;

public class Subscribing
{
    [Fact]
    public async Task can_subscribe_to_events()
    {
        using var umaClient = UmaClient.Connect(new UmaClientOptions().WithHost("localhost").WithPort(50051));
        var tag = $"subscribe-{Guid.NewGuid()}";
        var orderCreated = new OrderCreated(Guid.NewGuid(), 42m);
        var eventToAppend = new UmaEvent(
            nameof(OrderCreated),
            JsonSerializer.SerializeToUtf8Bytes(orderCreated),
            [tag]);

        var received = new TaskCompletionSource<SequencedUmaEvent>();
        var ct = TestContext.Current.CancellationToken;

        using var subscription = umaClient.Subscribe(
            UmaFilter.Where([nameof(OrderCreated)], [tag]),
            evt => received.TrySetResult(evt),
            ct);

        await umaClient.AppendAsync([eventToAppend], ct: ct);

        AssertReceivedOrderCreated(await received.Task, tag, orderCreated);
    }

    [Fact]
    public async Task can_subscribe_to_events_using_read_async()
    {
        using var umaClient = UmaClient.Connect(new UmaClientOptions().WithHost("localhost").WithPort(50051));
        var tag = $"subscribe-{Guid.NewGuid()}";
        var orderCreated = new OrderCreated(Guid.NewGuid(), 99m);
        var eventToAppend = new UmaEvent(
            nameof(OrderCreated),
            JsonSerializer.SerializeToUtf8Bytes(orderCreated),
            [tag]);

        var received = new TaskCompletionSource<SequencedUmaEvent>();
        var ct = TestContext.Current.CancellationToken;

        var subscription = umaClient.ReadAsync(
            UmaFilter.Where([nameof(OrderCreated)], [tag]).WithOptions(o => o.Subscribe = true),
            ct);

        _ = Task.Run(async () =>
        {
            await foreach (var batch in subscription)
            {
                foreach (var evt in batch.Events)
                    received.TrySetResult(evt);
            }
        }, ct);

        await umaClient.AppendAsync([eventToAppend], ct: ct);

        AssertReceivedOrderCreated(await received.Task, tag, orderCreated);
    }

    static void AssertReceivedOrderCreated(SequencedUmaEvent received, string tag, OrderCreated expected)
    {
        Assert.Equal(nameof(OrderCreated), received.Event.EventType);
        Assert.Contains(tag, received.Event.Tags ?? []);
        Assert.Equal(expected, JsonSerializer.Deserialize<OrderCreated>(received.Event.Data.ToArray()));
    }
}

File: Tests.Fsharp/Client/Subscribing.fs
module Tests.Client.Subscribing

open System
open System.Text.Json
open System.Threading
open System.Threading.Tasks
open FSharp.Control
open FsUnit.Xunit
open Xunit
open UmaDb.Client.ClientBuilder
open UmaDb.Client.Event
open UmaDb.Client.Query
open UmaDb.Client.Operations

type OrderCreated = { OrderId: Guid; Amount: decimal }

let private assertReceivedOrderCreated (received: SequencedUmaEvent) (tag: string) (expected: OrderCreated) =
    received.Event.EventType |> should equal "OrderCreated"
    received.Event.Tags |> Option.iter (fun tags -> tags |> should contain tag)
    let payload = JsonSerializer.Deserialize<OrderCreated>(received.Event.Data.ToArray())
    payload.OrderId |> should equal expected.OrderId
    payload.Amount |> should equal expected.Amount

[<Fact>]
let ``can subscribe to events using subscribeWithCallback`` () =
    task {
        use umaClient = connect "localhost" 50002 |> build
        let tag = $"subscribe-cb-{Guid.NewGuid()}"
        let orderCreated = { OrderId = Guid.NewGuid(); Amount = 42m }
        let eventToAppend =
            { EventType = "OrderCreated"
              Data = ReadOnlyMemory(JsonSerializer.SerializeToUtf8Bytes orderCreated )
              Tags = Some [ tag ]
              Id = None }

        let received = TaskCompletionSource<SequencedUmaEvent>()
        let ct = CancellationToken.None
        let query = [ { Types = [ "OrderCreated" ]; Tags = [ tag ] } ]

        use _subscription =
            subscribeWithCallback 
                umaClient 
                ct 
                query 
                (fun (evt, _) ->
                    task {
                        received.TrySetResult(evt) |> ignore
                    })

        let! _ = appendOperation [ eventToAppend ] |> append umaClient ct
        let! evt = received.Task
        assertReceivedOrderCreated evt tag orderCreated
    }

[<Fact>]
let ``can subscribe using read with subscribe option`` () =
    task {
        use umaClient = connect "localhost" 50002 |> build
        let tag = $"subscribe-async-{Guid.NewGuid()}"
        let orderCreated = { OrderId = Guid.NewGuid(); Amount = 99m }
        let eventToAppend =
            { EventType = "OrderCreated"
              Data = ReadOnlyMemory(JsonSerializer.SerializeToUtf8Bytes(orderCreated))
              Tags = Some [ tag ]
              Id = None }

        let received = TaskCompletionSource<SequencedUmaEvent>()
        let ct = CancellationToken.None
        let query = [ { Types = [ "OrderCreated" ]; Tags = [ tag ] } ]
        let options = QueryOptions.defaults |> QueryOptions.subscribe

        let _ =
            Task.Run(fun () ->
                task {
                    for evt in readWithOptions umaClient ct query options do
                        received.TrySetResult(evt) |> ignore
                } :> Task)

        let! _ = appendOperation [ eventToAppend ] |> append umaClient ct
        let! evt = received.Task
        assertReceivedOrderCreated evt tag orderCreated
    }
